pkg/apis/gcpfirewall/v1alpha1/types.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Protocol defines network protocols supported for GCP firewall.
+type Protocol string
+
+// CIDR defines a IP block.
+// TODO(b/186120065) Modify the validation to include IPv6 CIDRs with FW 3.0 support.
+// +kubebuilder:validation:Pattern=`^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(/(3[0-2]|2[0-9]|1[0-9]|[0-9]))?$`
+type CIDR string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolICMP is the ICMP protocol.
+	ProtocolICMP Protocol = "ICMP"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:shortName=gf
+
+// GCPFirewall describes a GCP firewall spec that can be used to configure GCE
+// firewalls. A GCPFirewallSpec will correspond 1:1 with a GCE firewall rule.
+type GCPFirewall struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired configuration for GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Spec GCPFirewallSpec `json:"spec,omitempty"`
+
+	// Status is the runtime status of this GCP firewall
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	Status GCPFirewallStatus `json:"status,omitempty"`
+}
+
+// GCPFirewallSpec provides the specification of a GCPFirewall.
+// The firewall rule apply to the cluster associated targets (network tags or
+// secure tags) which are deduced by the controller. As a result, the specified
+// rule applies to ALL nodes and pods in the cluster.
+type GCPFirewallSpec struct {
+	// Name of the firewall rule that's configured in GCP.
+	// For classic GCP firewalls, this field needs to be unique. For FW 3.0,
+	// this field will be in the rule description.
+	Name string `json:"name"`
```

Comment:
I am wondering if you could just reuse the name of the GCPFirewall object as that one is unique within the API server. Why do you need another name in the CR?

Reply:
This CR will have multiple instances in the API server. Just for ingress controller, there will be two: one for L7 firewall, another one for L4 firewall. GCP firewall controller manages all the firewall CRs.

Reply:
I agree with @freehan, we can remove this field now. We added this field as the previous version of the spec can define multiple GCE rules. 

@sugangli Can you clarify you comment?. Each GCE firewall that Ingress/ Service controllers wants to create is a separate GCPFirewall CR. So, there is a one-to-one map between GCPFirewall CR and the associated GCE firewall rule.

Reply:
I thought it was the name of the GCPFirewall CR. In that case, we can remove it. How about the name in the GCPFirewallStatus?

Reply:
I think we should keep the name in GCPFirewallStatus because GCPFirewall controller may choose take control of naming scheme in future

Reply:
Updated.